src/main/java/org/skriptlang/skript/bukkit/tags/TagType.java

```diff
@@ -0,0 +1,131 @@
+package org.skriptlang.skript.bukkit.tags;
+
+import ch.njol.skript.lang.Expression;
+import ch.njol.skript.lang.SkriptParser;
+import ch.njol.util.Kleenean;
+import org.bukkit.Keyed;
+import org.bukkit.Material;
+import org.bukkit.entity.EntityType;
+import org.jetbrains.annotations.Contract;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.UnmodifiableView;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * The type of a tag. Represents a category or context that the tags apply in.
+ * For example, {@link #ITEMS} tags apply to {@link Material}s, like {@link #BLOCKS}, but in item form rather than as
+ * placed blocks.
+ * <br>
+ * This class also contains a static registry of all tag types.
+ *
+ * @param <T> see type.
+ */
+public class TagType<T extends Keyed> {
+
+	private static final List<TagType<?>> REGISTERED_TAG_TYPES = new ArrayList<>();
+
+	public static final TagType<Material> ITEMS = new TagType<>("item", Material.class);
+	public static final TagType<Material> BLOCKS = new TagType<>("block", Material.class);
+	public static final TagType<EntityType> ENTITIES = new TagType<>("entity [type]", "entity type", EntityType.class);
+
+	static {
+		TagType.addType(ITEMS, BLOCKS, ENTITIES);
+	}
+
+	private final String pattern;
+	private final String toString;
+	private final Class<T> type;
+
+	/**
+	 * @param pattern The pattern to use when construction the selection Skript pattern.
+	 * @param type The class this tag type applies to.
+	 */
+	public TagType(String pattern, Class<T> type) {
```

Comment:
TagType is really only there to mirror the way minecraft tags are in different registries. It isn't really intended for custom tag types, just for future additions if minecraft/bukkit/paper add more. Do you think it'd be worth it to make it more focused on allowing custom types?

Reply:
It's not just for as doing custom tags, it's mainly for the API to filter selected tags and return as such without doing extra computations.

Reply:
I still don't really see the use case, can you give me an example? The BLOCKS tag type can't use Material methods because it acts on tags, not the contents of tags. I think getMatchingTags is plenty versatile for things like this.